### PR TITLE
Runtime changes to support removing direct reference from System.Runtime.WindowsRuntime to System.Private.CoreLib

### PR DESCRIPTION
**Pull request by jkoritzinsky**
- Remove Internal.Runtime.InteropServices.WindowsRuntime.ExceptionSupport (moved up to System.Runtime.WindowsRuntime in the corresponding CoreFX PR).
- Change `WindowsRuntimeResourceManagerBase` to `IWindowsRuntimeResourceManager` and use a reflection-based wrapper to call the methods on the instance created from `System.Runtime.WindowsRuntime` (which was already created through reflection). Keep a version of `WindowsRuntimeResourceManagerBase` that implements `IWindowsRuntimeResourceManager` for the `ENABLE_WINRT` path used by ProjectN so as to not require reactionary changes there.
- Make `System.Runtime.InteropServices.WindowsRuntime.WindowsRuntimeMarshal.ReportUnhandledError` public so it can be exposed by the corefx PR.

See dotnet/corefx#40692 for the corresponding changes in corefx.

### CODE REVIEW

**Comment by jkotas** [comment]
Why do we need to do this?

**Comment by jkoritzinsky** [comment]
@ericstj mentioned that we want to reduce the number of times System.Runtime.WindowsRuntime is built/copied because it depends on System.Private.CoreLib. He has more information about the reason for these changes.

**Comment by jkotas** [comment]
I do not think we should be doing any changes around this until we have a plan for how to evolve the WinRT support going to forward.

**Comment by jkotas** [comment]
> @ericstj mentioned that we want to reduce the number of times System.Runtime.WindowsRuntime is built/copied because it depends on System.Private.CoreLib. He has more information about the reason for these changes.

@ericstj Could you please shed some light on this?

I would expect that the way to reduce number of times System.Runtime.WindowsRuntime is built is to harvest all configurations from old packages, except netcoreapp-Windows_NT.

**Comment by ericstj** [comment]
We'd like to not set a precedent for adding another DLL to the package every time we ship a new netcoreapp version, and we don't want to count on the old version working against the new Private.CoreLib.

If we can make System.Runtime.WindowsRuntime.* behave as normal assemblies that only depend on public surface area exposed in references then we can target a minimum TFM and no longer grow the package.

Having the package behave partially inbox/out-of-box (by omitting the windows implementation) is undesireable as it means we have to freeze reference assemblies and will end up causing package growth over time if we ever want to update API / versioning.  I'd like to get this package to be completely inbox or completely out-of-box.  Completely inbox has been deemed undesireable since it implies a Windows.winmd on linux.

**Comment by jkotas** [comment]
> package growth over time if we ever want to update API / versioning.

If we ever add new APIs to this contract, it will very likely require matching runtime updates. The new APIs won't work on old runtimes.

**Comment by ericstj** [comment]
> If we ever add new APIs to this contract, it will very likely require matching runtime updates. The new APIs won't work on old runtimes.

Suppose we wanted to add Span APIs to buffer extensions?  I don't think any of those extension methods have runtime coupling.  Sure projected types do, but these assemblies aren't exclusively projected types and I don't expect us to agree to add any more projected types in the future.  I have a hard time believing we'd ever add runtime coupled surface area to these contracts, whereas I can come up with many cases for adding non-runtime coupled API to these.

**Comment by jkotas** [comment]
This .dll is dependent on by the runtime for WinRT projections (https://github.com/dotnet/coreclr/blob/dc86291b53793e2d2d7d7961195cf74471accc82/src/md/winmd/inc/adapter.h#L87). We have outstanding ask to rev the WinRT projections for .NET 5. We are working through the potential designs, but I think there is a good chance that this will require changes in the public surface coupled with the runtime.

File: src/System.Private.CoreLib/shared/Internal/Resources/WindowsRuntimeResourceManagerBase.cs [hunk 1]
```diff
@@ -2,32 +2,37 @@
...
 
+using System;
 using System.Globalization;
 
 namespace Internal.Resources
 {
-    // This is implemented in System.Runtime.WindowsRuntime as System.Resources.WindowsRuntimeResourceManager,
-    // allowing us to ask for a WinRT-specific ResourceManager.
-    public abstract class WindowsRuntimeResourceManagerBase
+    // This implementation wraps the WindowsRuntimeResourceManager implementation in ProjectN to adapt it to the new IWindowsRuntimeResourceManager interface
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/Internal/Resources/WindowsRuntimeResourceManagerBase.cs]: We are not longer syncing CoreLib with ProjectN. You do not need to worry about keeping ProjectN working.

**Comment by jeffschwMSFT** [comment]
@jkotas @ericstj my desire in asking @jkoritzinsky to do this work was only to help aid the nuget build/ship related maters.  Eg. every time we ship a .NET Core we need to rev/update this package.  My understanding is that if we get onto the public contract we can get off that treadmill (which is desirable).

The connection between this assembly and the runtime is deep and not changing.  It is possible that this will be part of our future looking plans, but I do not think that this current works will hurt that work.  The only way it may hamper us is if we deem it necessary to make API changes and we really do not desire to expose the new dependencies as public surface area.

Given we are solely focused on building against the contracts to help this package be less special, is there a concern with this work?  (AFAIK, we still need to validate with @ericstj that this work enables the desired outcome).

**Comment by jkotas** [comment]
> every time we ship a .NET Core we need to rev/update this package.

Do we really need to do that?

Let's say we do not add any new APIs to this contract in .NET 5 and we do not rev/update this package for .NET 5. Somebody makes .NET 5 app and references .NET Core 3.0 version of this package to get access to the S.R.WR APIs. What is not going to work?

> Given we are solely focused on building against the contracts to help this package be less special, is there a concern with this work?

I am worried that making the package completely out-of-box will enable scenarios that we do not want to support. Completely out-of-box packages (like System.Reflection.Metadata) can be mixed and matched with runtime versions, e.g. you would be able to mix .NET 6 version of the package and implementation with .NET 5 runtime. I do not think we want to sign up for that with S.R.WR given how special it is.

**Comment by jkoritzinsky** [comment]
One possible idea could be to move the implementations that currently live in System.Runtime.WindowsRuntime down to System.Runtime.InteropServices.WindowsRuntime (which is currently a facade) and add type forwards from System.Runtime.WindowsRuntime to System.Runtime.InteropServices.WindowsRuntime. Then we have a fully separate System.Runtime.WindowsRuntime assembly, a fully in-box System.Runtime.InteropServices.WindowsRuntime assembly, and no assemblies crossing the boundary of in-box/out-of-box.

**Comment by jeffschwMSFT** [comment]
> I am worried that making the package completely out-of-box will enable scenarios that we do not want to support

Completely agree.  This works does not make the current situation better or worse.

A complication with the current implementation is that this is parented under netstandard2.0 - but it already does not handle .NET Core 2.* which is netstandard compliant.

>> every time we ship a .NET Core we need to rev/update this package.
> Do we really need to do that?

My understanding is that the current policy is to ship it every time.  I chatted with @ericstj of changing the policy, but the policy is in place for a reasonable reason.

**Comment by jkotas** [comment]
> One possible idea could be to move the implementations that currently live in System.Runtime.WindowsRuntime down to System.Runtime.InteropServices.WindowsRuntime 

That has the problem of requiring Windows.winmd on Unix.

> the policy is in place for a reasonable reason.

Right, so we agree that this package will ship with every release, regardless of what we do here.

> A complication with the current implementation is that this is parented under netstandard2.0

The netstandard2.0 implementation is filled with PlatformNotSupported stubs. I cannot think about what we would do differently for netstandard2.0 implementation here.

> This works does not make the current situation better or worse.

This work makes the situation more complicated. We need to keep remembering that the whole history of the package.

I see that the latest package at https://www.nuget.org/packages/System.Runtime.WindowsRuntime/4.6.0-preview8.19405.3 has implementation at `runtimes\win\lib\netcoreapp3.0\System.Runtime.WindowsRuntime.dll`. Is that correct? I would expect it to be `_._` placeholder since the implementation should always come from the runtime. If we change the implementation to be placeholder here, would it address the build/ship matters?

**Comment by jkoritzinsky** [comment]
I'll close this for now.

**Pull request closed without merging**